py_zipkin/request_helpers.py

```diff
@@ -0,0 +1,147 @@
+        'sample_decision_only': False,
+    }
+    if headers.get('X-B3-Flags') == '1':
+        parsed['sampled_str'] = 'd'
+    # Unlike the single-header, "absent" and "present-but-empty-string" are
+    # both representable, and the spec only mentions Absent for encoding
+    # "Defer".  So that means we need to translate an empty-string to a
+    # non-empty, but invalid string.
+    if parsed['sampled_str'] == '':
+        parsed['sampled_str'] = 'INVALID'
+    parsed['sample_decision_only'] = (
+        all(parsed[k] is None for k in ('trace_id', 'span_id', 'parent_span_id'))
+        and parsed['sampled_str'])
+    return parsed
+
+
+def extract_zipkin_attrs_from_headers(headers, sample_rate=100.0,
+                                      use_128bit_trace_id=False):
+    """
+    Implements extraction of B3 headers per:
+        https://github.com/openzipkin/b3-propagation
+
+    The input headers can be any dict-like container that supports "in"
+    membership test and a .get() method that accepts a default value.
+
+    Returns a ZipkinAttrs instance or None
+    """
+    if 'b3' in headers:
+        parsed = _parse_single_header(headers['b3'])
+    else:
+        parsed = _parse_multi_header(headers)
+
+    # Handle the lone-sampling-decision case:
+    if parsed['sample_decision_only']:
+        # See if it's a valid encoding of (only) sample decision
+        if parsed['sampled_str'] in ('0', 'false'):
+            sample_rate = 0.0
+        elif parsed['sampled_str'] in ('1', 'true', 'd'):
+            sample_rate = 100.0
+        else:
+            # Nonsense gets a None
+            log.warning('Ignoring invalid sampling decision %r; '
+                        'requires "0", "1", or "d".', parsed['sampled_str'])
+            return None
+        attrs = create_attrs_for_span(
+            sample_rate=sample_rate,
+            use_128bit_trace_id=use_128bit_trace_id,
+            flags='1' if parsed['sampled_str'] == 'd' else '0',
+        )
+        return attrs
+
+    # Enforce requirement of having trace & span IDs
+    if not parsed['trace_id'] or not parsed['span_id']:
+        log.warning('Ignoring invalid or missing TraceId & SpanId:'
+                    '%r, %r', parsed['trace_id'], parsed['span_id'])
+        return None
+
+    # Handle sampling decision
+    if parsed['sampled_str']:
+        # We have 1==Accept, 0==Deny, d==Debug
+        if parsed['sampled_str'] in ('1', 'true', 'd'):
+            is_sampled = True
+        elif parsed['sampled_str'] in ('0', 'false'):
+            is_sampled = False
+        else:
+            # Not valid according to spec
+            log.warning('Invalid SamplingState %r; requires '
+                        '"0", "1", or "d".', parsed['sampled_str'])
+            return None
+    else:
+        # sample flag missing; means "Defer" and we're responsible for
+        # rolling fresh dice
+        is_sampled = _should_sample(sample_rate)
```

Comment:
I'd do this externally, especially as sampling like this is only one form of sampling. ex later it could be done via http parameters. having the sample rate code here is unnecessarily coupled.

Reply:
Well, the `ZipkinAttrs` are a namedtuple, which is a tuple, which is immutable.  If a `ZipkinAttrs` gets out of here with a `is_sampled` field that just needs to get filled in immediately, a whole new ZipkinAttrs would need to be created.

That just seems a bit byzantine to me...

I'll refer you to my example call-site above in another comment and ask again how that could be better?  I'm pretty happy with it.

Reply:
I'll leave this point up to @drolando as it is more about maintenance than anything else, and I don't maintain this particular code.